http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala

```diff
@@ -831,23 +848,53 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
     _ <- F.raiseUnless(hasWorkspacePermission)(ForbiddenError(userInfo.userEmail))
 
     allClusters <- KubernetesServiceDbQueries.listFullAppsByWorkspaceId(Some(workspaceId), Map.empty).transaction
-    apps = allClusters
-      .flatMap(_.nodepools)
-      .flatMap(n => n.apps)
 
-    nonDeletableApps = apps.filterNot(app => AppStatus.deletableStatuses.contains(app.status))
+    // need the cloudContext to delete the resources, get it from the KubernetesCluster here
+    // apps = List(App, CloudContext)
+    apps = allClusters.flatMap { cluster =>
+      val cloudContext = cluster.cloudContext
+      cluster.nodepools.flatMap { nodepool =>
+        nodepool.apps.map { app =>
+          (app, cloudContext)
+        }
+      }
+    }
+
+    nonDeletableApps = apps.filterNot(app => AppStatus.deletableStatuses.contains(app._1.status)).map(_._1)
 
     _ <- F
       .raiseError(DeleteAllAppsCannotBePerformed(workspaceId, nonDeletableApps, ctx.traceId))
       .whenA(!nonDeletableApps.isEmpty)
 
-    _ <- apps
-      .traverse { app =>
-        deleteAppV2Base(app, userInfo, workspaceId, deleteDisk)
-      }
+    workspaceApi <- wsmClientProvider.getWorkspaceApi(userInfo.accessToken.token)
+    attempt <- F.delay(workspaceApi.getWorkspace(workspaceId.value, IamRole.READER)).attempt
+
+    _ <- attempt match {
+      // if the workspace is found, delete the app normally
+      case Right(workspaceDesc) =>
+        apps
+          .traverse { app =>
+            deleteAppV2Base(app._1, app._2, userInfo, workspaceId, deleteDisk, workspaceDesc)
+          }
+      // if the workspace can't be found, delete the workspace records
+      case Left(_: NotFoundException) =>
```

Comment:
Same comment as above re triple checking the NotFoundException only corresponds to a 404

Reply:
we do already check workspace permission on line 848
`_ <- F.raiseUnless(hasWorkspacePermission)(ForbiddenError(userInfo.userEmail))`

but yes on the workspace swagger there's a separate `403 Permission denied` response